File: pipeline/detectors/c_implicit_declaration.py
"""
Detector for C implicit function declaration errors.
"""

import re
import typing as T
from pipeline.detectors.base import Detector
from pipeline.models import ErrorClue


class CImplicitDeclarationDetector(Detector):
    """
    Detect C implicit function declaration errors that suggest missing includes.

    Matches patterns like:
    - error: implicit declaration of function 'printf' [-Werror=implicit-function-declaration]
    - note: include '<stdio.h>' or provide a declaration of 'printf'
    """

    PATTERNS = {
        "missing_c_include": r"(?P<file_path>[a-zA-Z0-9_./\-]+\.c):\d+:\d+:\s+(?:error|warning):\s+implicit declaration of function\s+['\u2018](?P<function_name>[^'\u2019]+)['\u2019].*?note:\s+include\s+['\u2018]<(?P<suggested_include>[^>]+)>['\u2019]",
        "missing_c_function": r"(?P<file_path>[a-zA-Z0-9_./\-]+\.c):(?P<line_number>\d+):\d+:\s+(?:error|warning):\s+implicit declaration of function\s+['\u2018](?P<function_name>[^'\u2019]+)['\u2019]",
    }

    EXAMPLES = [
        (
            "test.c:5:5: error: implicit declaration of function 'printf'\nnote: include '<stdio.h>' or provide a declaration of 'printf'",
            {
                "clue_type": "missing_c_include",
                "confidence": 1.0,
                "context": {
                    "file_path": "test.c",
                    "function_name": "printf",
                    "suggested_include": "stdio.h",
                },
            },
        ),
        (
            "test.c:10:5: error: implicit declaration of function 'myFunction'",
            {
                "clue_type": "missing_c_function",
                "confidence": 1.0,
                "context": {
                    "file_path": "test.c",
                    "line_number": "10",
                    "function_name": "myFunction",
                },
            },
        ),
    ]

File: src/boil/detectors/c_implicit_declaration.go
// Detector for C implicit function declaration errors.
package detectors

// CImplicitDeclarationDetector detects C implicit function declaration errors
type CImplicitDeclarationDetector struct {
	*BaseDetector
}

func NewCImplicitDeclarationDetector() (*CImplicitDeclarationDetector, error) {
	base, err := NewBaseDetector(
		"CImplicitDeclarationDetector",
		100,
		map[string]string{
			"missing_c_include":  `(?P<file_path>[a-zA-Z0-9_./\-]+\.c):\d+:\d+:\s+(?:error|warning):\s+implicit declaration of function\s+[''](?P<function_name>[^'']+)[''].*?note:\s+include\s+['']<(?P<suggested_include>[^>]+)>['']`,
			"missing_c_function": `(?P<file_path>[a-zA-Z0-9_./\-]+\.c):(?P<line_number>\d+):\d+:\s+(?:error|warning):\s+implicit declaration of function\s+[''](?P<function_name>[^'']+)['']`,
		},
		[]DetectorExample{
			{
				Name:     "implicit_declaration_with_note",
				Input:    "main.c:10:5: warning: implicit declaration of function 'printf' [-Wimplicit-function-declaration] note: include '<stdio.h>'",
				ClueType: "missing_c_include",
				Context: map[string]string{
					"file_path":         "main.c",
					"function_name":     "printf",
					"suggested_include": "stdio.h",
				},
			},
			{
				Name:     "implicit_declaration_simple",
				Input:    "test.c:42:8: error: implicit declaration of function 'custom_func'",
				ClueType: "missing_c_function",
				Context: map[string]string{
					"file_path":     "test.c",
					"line_number":   "42",
					"function_name": "custom_func",
				},
			},
		},
	)
	if err != nil {
		return nil, err
	}
	return &CImplicitDeclarationDetector{base}, nil
}
